lightning-transaction-sync/src/esplora.rs

```diff
@@ -0,0 +1,376 @@
+			tx1.block_height.cmp(&tx2.block_height).then_with(|| tx1.pos.cmp(&tx2.pos))
+		});
+
+		Ok((confirmed_txs, unconfirmed_registered_txs, unspent_registered_outputs))
+	}
+
+	#[maybe_async]
+	fn get_confirmed_tx(
+		&self, txid: &Txid, expected_block_hash: Option<BlockHash>, known_block_height: Option<u32>,
+	) -> Result<Option<ConfirmedTx>, TxSyncError> {
+		if let Some(merkle_block) = maybe_await!(self.client.get_merkle_block(&txid))? {
+			let block_header = merkle_block.header;
+			let block_hash = block_header.block_hash();
+			if let Some(expected_block_hash) = expected_block_hash {
+				if expected_block_hash != block_hash {
+					log_trace!(self.logger, "Inconsistency: Tx {} expected in block {}, but is confirmed in {}", txid, expected_block_hash, block_hash);
+					return Err(TxSyncError::Inconsistency);
+				}
+			}
+
+			let mut matches = vec![*txid];
+			let mut indexes = Vec::new();
+			let _ = merkle_block.txn.extract_matches(&mut matches, &mut indexes);
+			debug_assert_eq!(indexes.len(), 1);
+			let pos = *indexes.get(0).ok_or(TxSyncError::Failed)? as usize;
+			if let Some(tx) = maybe_await!(self.client.get_tx(&txid))? {
+				if let Some(block_height) = known_block_height {
+					// We can take a shortcut here if a previous call already gave us the height.
+					return Ok(Some(ConfirmedTx { tx, block_header, pos, block_height }));
+				}
+
+				let block_status = maybe_await!(self.client.get_block_status(&block_hash))?;
+				if let Some(block_height) = block_status.height {
+					return Ok(Some(ConfirmedTx { tx, block_header, pos, block_height }));
+				} else {
+					// If any previously-confirmed block suddenly is no longer confirmed, we found
+					// an inconsistency and should start over.
+					log_trace!(self.logger, "Inconsistency: Tx {} was unconfirmed during syncing.", txid);
+					return Err(TxSyncError::Inconsistency);
+				}
+			}
+		}
+		Ok(None)
+	}
+
+	#[maybe_async]
+	fn sync_unconfirmed_transactions(
+		&self, confirmables: &Vec<&(dyn Confirm + Sync + Send)>,
+	) -> Result<(), TxSyncError> {
+		// Query the interface for relevant txids and check whether the relevant blocks are still
+		// in the best chain, mark them unconfirmed otherwise. If the transactions have been
+		// reconfirmed in another block, we'll confirm them in the next sync iteration.
+		let relevant_txids = confirmables
+			.iter()
+			.flat_map(|c| c.get_relevant_txids())
+			.collect::<HashSet<(Txid, Option<BlockHash>)>>();
+
+		let mut locked_watched_transactions = self.watched_transactions.lock().unwrap();
+		for (txid, block_hash_opt) in relevant_txids {
+			if let Some(block_hash) = block_hash_opt {
+				let block_status = maybe_await!(self.client.get_block_status(&block_hash))?;
+				if block_status.in_best_chain {
+					// Skip if the block in question is still confirmed.
+					continue;
+				}
+			}
+
+			for c in confirmables {
+				c.transaction_unconfirmed(&txid);
+			}
+
+			locked_watched_transactions.insert(txid);
```

Comment:
I now included this essentially as a hotfix for the latest feedback from https://github.com/lightningdevkit/ldk-lite/pull/9. However, I'm still not sure if we shouldn't rather go either with the "simply always confirm everything" or the "re-register outputs via `load_outputs_to_watch` upon start of the sync round" approaches.

Reply:
putting this here causes it so confirmed txs are never added to `watched_transactions` and thus never confirmed

Reply:
Huh, I'm not sure why they never would be confirmed?

Generally all transactions will be added to `locked_watched_transactions` when the `Filter` queues are processed via `process_queues()`. Any transaction returned by `get_relevant_txids()` only needs to be monitored for re- and unconfirmations, in which case we call `transaction_confirmed` and re-add them to the list to monitor for confirmation, i.e., `watched_transactions`. Am I missing something?

Reply:
Haven't debugged it yet but we're using a copy-paste of this in mutiny and transactions would never get confirmed unless we added txs that were confirmed here

Reply:
Mh, I'm  still not clear why that would happen when only re-adding the unconfirmed transactions. 

I now cleaned up the approach a bit with https://github.com/lightningdevkit/rust-lightning/pull/1870/commits/5d94cb65c447ed4e607fc90006ba1aed9e3955fe. Could you check whether this mitigates what you saw before? 

Btw. I now also migrated to `future::lock::Mutex`. Does this allow you to re-use the create instead of copy-pasting?